### DISCUSSION TITLE
update many to many relationship by hybrid_property

### DISCUSSION THREAD

**Discussion by finswimmer** [sqlalchemy/sqlalchemy#7488] [Usage Questions]
Hello,

I have two ORM models `Person` and `Manager`. A person can have one manager, but a manager can be related to multiple persons. So these to models are linked via a third model `PersonManagerMap`.

The `Person` model contains a  hybrid_property that reflects the managers name (which is unique). I'm looking for a solution to change the relationship by setting the managers name in the `Person` model.

This works for the case the instance is attached to a session:

```py
class PersonManagerMap(Base):
    __tablename__ = "person_manager_maps"

    id = Column(Integer, primary_key=True)
    emp_id = Column(ForeignKey("persons.emp_id"), unique=True, nullable=False)
    man_id = Column(ForeignKey("managers.man_id"), nullable=False)


class Person(Base):
    __tablename__ = "persons"

    emp_id = Column(Integer, primary_key=True)
    person_name = Column(String(200), unique=True, nullable=False)

    manager: "Manager" = relationship(
        "Manager", secondary=PersonManagerMap.__table__, uselist=False
    )

    @hybrid_property
    def manager_name(self) -> str:
        return self.manager.manager_name

    @manager_name.setter
    def manager_name(self, value: str) -> None:
        self.manager = object_session(self).query(Manager).filter_by(manager_name=value).one()


class Manager(Base):
    __tablename__ = "managers"

    man_id = Column(Integer, primary_key=True)
    manager_name = Column(String(200), unique=True, nullable=False)
```

Is there a way to get rid of the `object_session` to use the setter for instances that are not yet attached to a session?

I tried:

```py
    @manager_name.setter
    def manager_name(self, value: str) -> None:
        self.manager = (
            select(Manager).where(Manager.manager_name == value).scalar_subquery()
        )
```

But this results in `AttributeError: 'Select' object has no attribute '_sa_instance_state'`.

Any hints for me?

fin swimmer

**Comment by gordthompson** [comment]
> A person can have one manager, but a manager can be related to multiple persons. So these to models are linked via a third model `PersonManagerMap`.

So that is a one-to-many relationship and an association table/object is not necessary. All you need is a foreign key in Person that points to their Manager.

**Reply by finswimmer** [reply]
> All you need is a foreign key in Person that points to their Manager.

You are right, that this is the correct solution for the concrete example. 

However, the example is just a minimal example to show the technical challenge.

**Comment by zzzeek** [comment]
you wouldn't use a hybrid property in this way, just a normal Python `@property` should be used as this attribute has no ability to be used as a query element.

second, Person.manager looks like a many-to-one, so it's not really appropriate to have an association table here.  Using uselist=False with a many-to-many relationship doesn't make any sense.   As Gord says, Person should have a simple "manager" foreign key on it.

thirdly, you want to set some_person().manager_name = "some manager" and that will do a lookup of the Manager with that name and assign it.  object_session() is the way to go there, since you have no choice but to look up this name in the database.  There are other ways to do this recipe, where you would resolve the Manager object at session flush time, but that would be more complicated to implement and i dont see what advantage that gets.
